[LoopVectorize] Vectorize select-cmp reduction pattern for increasing integer induction variable

Mel-Chen opened the pull request:
Consider the following loop:
```
  int rdx = init;
  for (int i = 0; i < n; ++i)
    rdx = (a[i] > b[i]) ? i : rdx;
```
We can vectorize this loop if `i` is an increasing induction variable.
The final reduced value will be the maximum of `i` that the condition
`a[i] > b[i]` is satisfied, or the start value `init`.

This patch added new RecurKind enums - IFindLastIV and FFindLastIV.

Mel-Chen committed:
[LoopVectorize] Vectorize select-cmp reduction pattern for increasing
integer induction variable

Consider the following loop:

  int rdx = init;
  for (int i = 0; i < n; ++i)
    rdx = (a[i] > b[i]) ? i : rdx;

We can vectorize this loop if `i` is an increasing induction variable.
The final reduced value will be the maximum of `i` that the condition
`a[i] > b[i]` is satisfied, or the start value `init`.

This patch added new RecurKind enums - IFindLastIV and FFindLastIV.

Mel-Chen committed:
Refactor the identification method for FindLastIV pattern.

This patch applys range analysis. It will exclude cases where the
range of induction variable cannot be fully contained within

  [<sentinel value> + 1, <minimum value of recurrence type>)

This approach also handles truncated induction variable cases well.

alexey-bataev commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -662,6 +664,96 @@ RecurrenceDescriptor::isAnyOfPattern(Loop *Loop, PHINode *OrigPhi,
+                         m_Value())))
+    return InstDesc(false, I);
+
+  SelectInst *SI = cast<SelectInst>(I);
```
```suggestion
  auto *SI = cast<SelectInst>(I);
```

alexey-bataev commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -662,6 +664,96 @@ RecurrenceDescriptor::isAnyOfPattern(Loop *Loop, PHINode *OrigPhi,
+  else
+    return InstDesc(false, I);
+
+  auto IsIncreasingLoopInduction = [&SE, &Loop](Value *V) {
```
```suggestion
  auto IsIncreasingLoopInduction = [&](Value *V) {
```

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8973,8 +8980,10 @@ void LoopVectorizationPlanner::adjustRecipesForReductions(
     RecurKind Kind = RdxDesc.getRecurrenceKind();
-    assert(!RecurrenceDescriptor::isAnyOfRecurrenceKind(Kind) &&
-           "AnyOf reductions are not allowed for in-loop reductions");
+    assert(
+        (!RecurrenceDescriptor::isAnyOfRecurrenceKind(Kind) &&
+         !RecurrenceDescriptor::isFindLastIVRecurrenceKind(Kind)) &&
+        "AnyOf and FindLast reductions are not allowed for in-loop reductions");
```
```suggestion
    assert(
        !RecurrenceDescriptor::isAnyOfRecurrenceKind(Kind) &&
         !RecurrenceDescriptor::isFindLastIVRecurrenceKind(Kind) &&
        "AnyOf and FindLast reductions are not allowed for in-loop reductions");
```

ShivaChen commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -662,6 +664,96 @@ RecurrenceDescriptor::isAnyOfPattern(Loop *Loop, PHINode *OrigPhi,
+    // virtual OR reduction.
+    const APInt Sentinel = APInt::getSignedMinValue(NumBits);
+    const ConstantRange ValidRange = ConstantRange::getNonEmpty(
+        Sentinel + 1, APInt::getSignedMinValue(NumBits));
```
Could you explain more about why picking SignedMin as Upper and how it effect the range comparison?

Mel-Chen replied:
Sure. 
This is because the ConstantRange::getNonEmpty(L, U) actually produces the range [L, U). If calling ConstantRange::getNonEmpty(SignedMin + 1, SignedMin), for integer type, [SignedMin + 1, SignedMin) is equivalent to [SignedMin + 1, SignedMax]. 
This part has been confusing for many people, including myself when I was initially learning about ConstantRange. I am currently considering change it to:
```
// Valid range = Full set - Sentinel 
const ConstantRange ValidRange = 
    ConstantRange::getFull(NumBits).difference(ConstantRange(Sentinel));
```
Please let me know if you think this representation would be better.

ShivaChen replied:
Yes, the new representataion seems express well. 
Thanks for your kind explaination. :-)

artagnon commented on llvm/test/Transforms/LoopVectorize/iv-select-cmp.ll:
```diff
@@ -1,10 +1,193 @@
-; RUN: opt -passes=loop-vectorize -force-vector-interleave=4 -force-vector-width=4 -S < %s | FileCheck %s --check-prefix=CHECK
-; RUN: opt -passes=loop-vectorize -force-vector-interleave=4 -force-vector-width=1 -S < %s | FileCheck %s --check-prefix=CHECK
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 3
+; RUN: opt -passes=loop-vectorize -force-vector-interleave=1 -force-vector-width=4 -S < %s | FileCheck %s --check-prefix=CHECK-VF4IC1 --check-prefix=CHECK
```
`FileCheck %s --check-prefixes=CHECK,CHECK-VF4IC1`.

Mel-Chen replied:
74662c0e26c1a2d8804082f40ddcf8872bcdb9b4
Done, thank you.

artagnon commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -662,6 +664,96 @@ RecurrenceDescriptor::isAnyOfPattern(Loop *Loop, PHINode *OrigPhi,
+    // virtual OR reduction.
+    const APInt Sentinel = APInt::getSignedMinValue(NumBits);
+    const ConstantRange ValidRange = ConstantRange::getNonEmpty(
+        Sentinel + 1, APInt::getSignedMinValue(NumBits));
```
`ConstantRange::getNonEmpty(Sentinel + 1, Sentinel)`.

artagnon reviewed (commented):
Looks good, apart from minor nits.

Mel-Chen committed:
Replace SelectInst with auto

Comment from Alexey

Co-authored-by: Alexey Bataev <5361294+alexey-bataev@users.noreply.github.com>

Mel-Chen committed:
Drop parentheses

comment from Alexey

Co-authored-by: Alexey Bataev <5361294+alexey-bataev@users.noreply.github.com>

Mel-Chen committed:
Modified lambda to capture all external variables by reference

Co-authored-by: Alexey Bataev <5361294+alexey-bataev@users.noreply.github.com>

Mel-Chen committed:
Fix format

Mel-Chen committed:
Clean up comments

Mel-Chen committed:
Refine the expression of valid range

ShivaChen commented:
Although IFindLastIV can be fixed to FFindLastIV when parsing instructions, could we reject IFindLastIV with FCmpInst and AddReductionVar for FFindLastIV explicitily?

Mel-Chen committed:
Fix the warning caused by unused parameter

Mel-Chen committed:
Fix typo

Mel-Chen committed:
Clean comment in test case

Mel-Chen commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -662,6 +664,96 @@ RecurrenceDescriptor::isAnyOfPattern(Loop *Loop, PHINode *OrigPhi,
+
+    const ConstantRange IVRange = SE->getSignedRange(AR);
+    unsigned NumBits = Ty->getIntegerBitWidth();
+    // Keep the minmum value of the recurrence type as the sentinel value.
```
Fix, thanks.

Mel-Chen commented on llvm/test/Transforms/LoopVectorize/iv-select-cmp-trunc.ll:
```diff
@@ -1,6 +1,7 @@
+; RUN: opt -passes=loop-vectorize -force-vector-interleave=4 -force-vector-width=4 -S < %s | FileCheck %s --check-prefix=CHECK-VF4IC4 --check-prefix=CHECK
+; RUN: opt -passes=loop-vectorize -force-vector-interleave=4 -force-vector-width=1 -S < %s | FileCheck %s --check-prefix=CHECK-VF1IC4 --check-prefix=CHECK
 
 ; This test can theoretically be vectorized without a runtime-check, by
```
Updated.

Mel-Chen commented:
> Although IFindLastIV can be fixed to FFindLastIV when parsing instructions, could we reject IFindLastIV with FCmpInst and AddReductionVar for FFindLastIV explicitily?

This is possible, but it may cause going through the pattern identification process one more time.

Mel-Chen commented:
@ayalz @fhahn @david-arm Ping.

david-arm reviewed (commented):
Thanks for this patch - it looks like a useful addition. I'm still looking through all the tests, but I thought I'd leave some comments I have so far!

Mel-Chen commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -900,6 +993,11 @@ bool RecurrenceDescriptor::isReductionPHI(PHINode *Phi, Loop *TheLoop,
                       << *Phi << "\n");
     return true;
   }
+  if (AddReductionVar(Phi, RecurKind::IFindLastIV, TheLoop, FMF, RedDes, DB, AC,
```
Good question. We don't need to call `AddReductionVar` again for FFindLastIV.
This is because at the end of function `isFindLastIVPattern`, IFindLastIV can be transformed into FFindLastIV if the predicate is an fcmp instruction.

Mel-Chen replied:
It looks like https://github.com/llvm/llvm-project/pull/118393 is attempting to improve this issue.

Mel-Chen commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -660,6 +662,96 @@ RecurrenceDescriptor::isAnyOfPattern(Loop *Loop, PHINode *OrigPhi,
+    // TODO: This range restriction can be lifted by adding an additional
+    // virtual OR reduction.
+    const APInt Sentinel = APInt::getSignedMinValue(NumBits);
+    const ConstantRange ValidRange =
```
Your answer is correct. :)
Actually, that's  my initial version. The current version has been modified based on previous comments. https://github.com/llvm/llvm-project/pull/67812/commits/2466df5a17a127d33bc1c29193ae4f8a7f80856b

Mel-Chen replied:
8d7a3a0168844b10feaad07990dc4f3827cdaa20
91d7478b1a7b48c7dc3863c2eb6bb4ea64fbe04f
It's just that I haven't made the change yet. It's done now. :)

Mel-Chen committed:
Fix SE pass

Mel-Chen committed:
Add TODO comment for the decreasing induction

Mel-Chen commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -660,6 +662,96 @@ RecurrenceDescriptor::isAnyOfPattern(Loop *Loop, PHINode *OrigPhi,
+    return InstDesc(false, I);
+
+  auto IsIncreasingLoopInduction = [&](Value *V) {
+    if (!SE)
```
952b5df5f394e5e7c6f91d7849e2cc3040b895d1
Nice suggestion. I pass SE by reference directly. Thanks.

Mel-Chen commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -660,6 +662,96 @@ RecurrenceDescriptor::isAnyOfPattern(Loop *Loop, PHINode *OrigPhi,
+      return false;
+
+    const SCEV *Step = AR->getStepRecurrence(*SE);
+    if (!SE->isKnownPositive(Step))
```
ec9217d8aec5d01da56751701ae57eb8ee72e415
Add the TODO for the decreasing induction.

david-arm commented on llvm/test/Transforms/LoopVectorize/select-min-index.ll:
Maybe just delete this file, since it looks practically the same as llvm/test/Transforms/LoopVectorize/iv-select-cmp-no-wrap.ll and doesn't seem to offer any extra value?

Mel-Chen replied:
No, we must retain `llvm/test/Transforms/LoopVectorize/select-min-index.ll`. `select-min-index.ll` demonstrates another semantics: minmax with index. This semantics is similar to FindLastIV but different.
```
int min = start_min;
int idx = start_idx;
for (int i = 0; i < n; i++)
  if (min > a[i]) {
    min = a[i];
    idx = i;
  }

// live-out: idx
```

david-arm reviewed (commented):
I've almost finished reviewing! Still have some tests to work through, but I thought I'd leave the comments I have so far. It looks like you've done a thorough job at testing all the edge cases though.

david-arm reviewed (commented):
Hi, just one last comment. I've finished reviewing now!

Mel-Chen committed:
Refine comments

Mel-Chen committed:
Refine debug dump

Mel-Chen commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -660,6 +662,94 @@ RecurrenceDescriptor::isAnyOfPattern(Loop *Loop, PHINode *OrigPhi,
+RecurrenceDescriptor::isFindLastIVPattern(PHINode *OrigPhi, Instruction *I,
+                                          ScalarEvolution &SE) {
+  // Only match select with single use cmp condition.
+  // TODO: Only handle single use for now.
```
6887d2e92a3af165870ae6027edeb0cb152f0983

Mel-Chen commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -660,6 +662,96 @@ RecurrenceDescriptor::isAnyOfPattern(Loop *Loop, PHINode *OrigPhi,
+    // Keep the minimum value of the recurrence type as the sentinel value.
+    // The maximum acceptable range for the increasing induction variable,
+    // called the valid range, will be defined as
+    //   [<sentinel value> + 1, SignedMin(<recurrence type>))
```
6887d2e92a3af165870ae6027edeb0cb152f0983
Looks better now, thank you.

Mel-Chen commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -900,6 +991,11 @@ bool RecurrenceDescriptor::isReductionPHI(PHINode *Phi, Loop *TheLoop,
   }
+  if (AddReductionVar(Phi, RecurKind::IFindLastIV, TheLoop, FMF, RedDes, DB, AC,
+                      DT, SE)) {
+    LLVM_DEBUG(dbgs() << "Found a FindLastIV reduction PHI." << *Phi << "\n");
```
9a0bf28b07b38dd8d3c934fa6fe1229c9306f188
Sure, I agree that explicitly indicating IFindLast or FFindLast is better. 
The new modification directly determines whether to print IFindLast or FFindLast based on the final result. This way, there is no need to call `AddReductionVar` twice, and the correct result can be indicated. 
If you feel that detecting IFindLast and FFindLast separately would be clearer, please let me know.

Mel-Chen commented on llvm/test/Transforms/LoopVectorize/iv-select-cmp.ll:
```diff
@@ -1,10 +1,193 @@
+; RUN: opt -passes=loop-vectorize -force-vector-interleave=4 -force-vector-width=4 -S < %s | FileCheck %s --check-prefix=CHECK-VF4IC4 --check-prefix=CHECK
+; RUN: opt -passes=loop-vectorize -force-vector-interleave=4 -force-vector-width=1 -S < %s | FileCheck %s --check-prefix=CHECK-VF1IC4 --check-prefix=CHECK
 
 define i64 @select_icmp_const_1(ptr %a, i64 %n) {
```
They are indeed quite similar. iv-select-cmp.ll is for testing IR generation, including the case of UF > 1. iv-select-cmp-no-wrap.ll is for testing whether vectorization is legal. 
My suggestion is to remove `select_icmp_nuw_nsw` from iv-select-cmp-no-wrap.ll, and retain `@select_icmp_const_1` and `@select_icmp_const_2`.
What do you think?

Mel-Chen replied:
@david-arm Ping.
Regarding my test file proposal, what do you think?

Mel-Chen committed:
Refine comments

Mel-Chen committed:
Revert "Refine the expression of valid range"

This reverts commit 4bd1df23a9cb944b00d85e958ce8a6a051439def.

Mel-Chen committed:
Refine the expression of valid range

Mel-Chen committed:
Capture LHS and RHS of select instruction by match.

Mel-Chen commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -660,6 +662,94 @@ RecurrenceDescriptor::isAnyOfPattern(Loop *Loop, PHINode *OrigPhi,
+    return ValidRange.contains(IVRange);
+  };
+
+  // We are looking for selects of the form:
```
f19afa46b79616fc3163c22e265ef30da9976154
Perhaps this modification make it better?

Mel-Chen commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -660,6 +662,94 @@ RecurrenceDescriptor::isAnyOfPattern(Loop *Loop, PHINode *OrigPhi,
+  auto *SI = cast<SelectInst>(I);
+  Value *NonRdxPhi = nullptr;
+
+  if (OrigPhi == dyn_cast<PHINode>(SI->getTrueValue()))
```
4504018c14ef3cdee535c4351235ec05c4f6c03f
Good suggestion, thanks!

Mel-Chen commented:
@ayalz @fhahn Ping.
Could you clearly specify the shortcomings of this PR that prevent it from being approved?
For example, do I need to complete the FindLast reduction and not only handle the induction variable, or any other reasons? This could help speed up the progress of this PR. Thank you.

Mel-Chen committed:
Rebase, and update test cases

Mel-Chen committed:
Restrict FindLastIV idiom to single-use reduction phi.

As the following case, the complex semantics of FindLastIV vectorization
are not yet supported:
```
for.header:
  %indvars = phi 0, %indvars.next
  %rdx.phi = phi 0, %rdx.phi.next
  br %pred, label %bb0, label %bb1

bb0:
  ...
  %select.bb0 = select %cmp0, %rdx.phi, %indvars
  br label %for.inc

bb1:
  ...
  %select.bb1 = select %cmp1, %rdx.phi, %indvars
   br label %for.inc

for.inc:
  ...
  %rdx.phi.next = phi %select.bb0, %select.bb1
  %indvars.next = add nuw nsw i64 %indvars, 1
  br %exitcond, label %for.end, label %for.body

for.end:
 ; external use of %rdx.phi.next
```
This patch bails out unsupported idiom during the reduction
identification phase, to prevent internal compiler error.

Mel-Chen committed:
Remove unused variable

Mel-Chen commented:
Rebased, please take a look.
Additionally, we have encountered a new case after switch vectorization was enabled 54b624c229852eb33c200bc24bf267a5e5f6be4d.
The same issue might also occur with AnyOf reduction, although we haven't encountered it yet.

Mel-Chen committed:
Refine pattern matcher

Mel-Chen committed:
Refine lit test checker format

Mel-Chen commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -659,6 +661,100 @@ RecurrenceDescriptor::isAnyOfPattern(Loop *Loop, PHINode *OrigPhi,
+  else
+    return InstDesc(false, I);
+
+  auto IsIncreasingLoopInduction = [&](Value *V) {
```
Your comment suggests removing the lambda function `IsIncreasingLoopInduction` and replacing it with a direct return statements, if I understand correctly. 
This is indeed possible, but I worry that it could reduce the readability, making it difficult for developers to understand that this code section is confirming whether an induction variable is increasing.
Could you suggest a better approach that balances readability?

Mel-Chen commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -659,6 +661,100 @@ RecurrenceDescriptor::isAnyOfPattern(Loop *Loop, PHINode *OrigPhi,
+    return InstDesc(false, I);
+
+  Value *NonRdxPhi = nullptr;
+  if (OrigPhi == dyn_cast<PHINode>(TrueVal))
```
d7d466130e7144a309c62af32276b11716f2cb37
Thank you for pointing this issue out! I also melted it into the pattern matcher above.
It looks more elegant now.

Mel-Chen commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -659,6 +661,100 @@ RecurrenceDescriptor::isAnyOfPattern(Loop *Loop, PHINode *OrigPhi,
+    // where <sentinel value> is SignedMin(<recurrence type>)
+    // TODO: This range restriction can be lifted by adding an additional
+    // virtual OR reduction.
+    const APInt Sentinel = APInt::getSignedMinValue(NumBits);
```
I tried to create a concrete example based on your description:
C source
```
unsigned int foo(unsigned int start, int *a, unsigned int n) {
    unsigned int rdx = start;
    for (unsigned int i = 0; i <= 2147483648; i++) {
        rdx = a[i] > 3 ? i : rdx;
    }
    return rdx;
}
```
LLVM IR
```
; Function Attrs: nofree norecurse nosync nounwind memory(argmem: read) vscale_range(8,1024)
define dso_local signext i32 @foo(i32 noundef signext %start, ptr nocapture noundef readonly %a, i32 noundef signext %n) local_unnamed_addr #0 {
entry:
  br label %for.body

for.cond.cleanup:                                 ; preds = %for.body
  %cond.lcssa = phi i32 [ %cond, %for.body ]
  ret i32 %cond.lcssa

for.body:                                         ; preds = %entry, %for.body
  %indvars.iv = phi i64 [ 0, %entry ], [ %indvars.iv.next, %for.body ]
  %rdx.08 = phi i32 [ %start, %entry ], [ %cond, %for.body ]
  %arrayidx = getelementptr inbounds nuw i32, ptr %a, i64 %indvars.iv
  %0 = load i32, ptr %arrayidx, align 4, !tbaa !6
  %cmp2 = icmp sgt i32 %0, 3
  %1 = trunc nuw i64 %indvars.iv to i32
  %cond = select i1 %cmp2, i32 %1, i32 %rdx.08
  %indvars.iv.next = add nuw nsw i64 %indvars.iv, 1
  %exitcond.not = icmp eq i64 %indvars.iv.next, 2147483649
  br i1 %exitcond.not, label %for.cond.cleanup, label %for.body, !llvm.loop !10
}
```
From my side, it seems that this case has not been vectorized.
```
LV: FindLastIV valid range is [-2147483647,-2147483648), and the signed range of {0,+,1}<%for.body> is [0,-2147483647)
```

Would you like me to add this test to the test file if the example I provided aligns with the scenario you are concerned about?

fhahn replied:
Would be great if you could add this as a test case.

Mel-Chen replied:
There is a same case `@not_vectorized_select_fcmp_invalid_const_ub` in `llvm/test/Transforms/LoopVectorize/iv-select-cmp-trunc.ll`.

Mel-Chen commented:
Ping

Mel-Chen commented:
Ping

alexey-bataev commented on llvm/test/Transforms/LoopVectorize/iv-select-cmp-no-wrap.ll:
```diff
@@ -1,8 +1,58 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 3
```
Please, update the checks here in a separate patch and then update it here

Mel-Chen replied:
https://github.com/Mel-Chen/llvm-project/commit/0acc18c6cf803b61d5950a88f964f7b4d82db341
Are you mentioning something like this?

Mel-Chen replied:
@alexey-bataev 
Could you confirm if my understanding is correct so that I can rebase?

Mel-Chen replied:
#117560

Mel-Chen committed:
Remove unused CHECK

huntergr-arm reviewed (approved):
I think this is in good shape. There's some overlap with the CSA PR, but this patch should generate more efficient code on many targets due to the narrower focus on induction values alone. We may be able to do some refactoring later, but that shouldn't hold this up further.

(Sorry it took me so long to get back to this and finish off the review)

fhahn commented on llvm/include/llvm/Analysis/IVDescriptors.h:
```diff
@@ -236,10 +253,26 @@ class RecurrenceDescriptor {
 
+  /// Returns the sentinel value used to replace the start value.
+  Value *getSentinelValue() const {
+    if (isFindLastIVRecurrenceKind(Kind)) {
```
Should this only ever be called for FindLastIV recurrences? Can it be na assert?

Mel-Chen replied:
Yes, it should be assertion.
9059eec7bc59ee4af1fa00ea0c66ac7145a5023e

fhahn commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -658,6 +660,93 @@ RecurrenceDescriptor::isAnyOfPattern(Loop *Loop, PHINode *OrigPhi,
+    LLVM_DEBUG(dbgs() << "LV: FindLastIV valid range is " << ValidRange
+                      << ", and the signed range of " << *AR << " is "
+                      << IVRange << "\n");
+    return ValidRange.contains(IVRange);
```
IIUC this ensures that the IV doesn't wrap, would be good to add a comment.

Mel-Chen replied:
fb67dfd5031cb0bd7e73c865c5d6b454e18d6a5a

fhahn commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -658,6 +660,93 @@ RecurrenceDescriptor::isAnyOfPattern(Loop *Loop, PHINode *OrigPhi,
+      return false;
+
+    auto *AR = dyn_cast<SCEVAddRecExpr>(SE.getSCEV(V));
+    if (!AR)
```
Would probably also be good to check if the AddRec is for the right loop and not for an outer loop? If is not in the current loop, it will just be a regular select of integer reduction, but would still be good to explicitly check (or assert if possible). Also would be good to make sure there's a test case where the select op is an IV from an outer loop.

Mel-Chen replied:
You’ve raised a great point! 
Although this issue hasn’t surfaced due to the recognition order of AnyOf and FindLastIV, I agree that adding a loop check is safer.
ee12efbfe6987ff1ad5de829f3c695e7fa70ebad
And please take a look the pre-commit test for nested loop: #118519

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -5106,8 +5106,9 @@ LoopVectorizationCostModel::selectInterleaveCount(ElementCount VF,
-              RdxDesc.getRecurrenceKind());
+          RecurKind RK = RdxDesc.getRecurrenceKind();
+          return RecurrenceDescriptor::isAnyOfRecurrenceKind(RK) ||
+                 RecurrenceDescriptor::isFindLastIVRecurrenceKind(RK);
```
needs test without forced interleave count ?

fhahn reviewed (commented):
Thanks for following up on all the comments, the last remaining thing to clarify from my side is what costs we currently assign to the new reduction kinds? 

We should probably have a few target-dependent tests that check cost-modeling.

Mel-Chen committed:
Refine the comment for getSentinelValue

Mel-Chen committed:
Replace if-condition with assert

Mel-Chen committed:
Add comment for wrap check

Mel-Chen committed:
Add check for ensure the loop of SCEVAddRec is the same as the loop of reduction phi

Mel-Chen commented on llvm/include/llvm/Analysis/IVDescriptors.h:
```diff
@@ -236,10 +253,26 @@ class RecurrenceDescriptor {
   /// actual type of the Phi if the recurrence has been type-promoted.
   Type *getRecurrenceType() const { return RecurrenceType; }
 
+  /// Returns the sentinel value used to replace the start value.
```
Sure, thank you!
5aa56de94e5639d4c9e7ea85d9e831f4d8467a3d

Mel-Chen commented:
> Thanks for following up on all the comments, the last remaining thing to clarify from my side is what costs we currently assign to the new reduction kinds?
> 
> We should probably have a few target-dependent tests that check cost-modeling.

Is the cost of VPInstruction::ComputeReductionResult currently accounted for? FindLastIV should only require additional computation in the middle block, similar to AnyOf.

Mel-Chen commented:
Rebase, and ping @fhahn. Do you have any questions and comments for this patch?

fhahn reviewed (approved):
LGTM, thanks!

Mel-Chen commented:
Thanks @fhahn @huntergr-arm :)

fhahn commented:
It looks like the patch didn't correctly handle epilogue vectorization and caused a mis-compile in SPEC2017 on AArch64.

I pushed 0e528ac404e13ed2d952a2d83aaf8383293c851e to fix the issue and hopefully avoid a revert

Mel-Chen commented:
> It looks like the patch didn't correctly handle epilogue vectorization and caused a mis-compile in SPEC2017 on AArch64.
> 
> I pushed [0e528ac](https://github.com/llvm/llvm-project/commit/0e528ac404e13ed2d952a2d83aaf8383293c851e) to fix the issue and hopefully avoid a revert

Thank you! This is something I didn't notice before.
I am going to check IR for epilogue vectorization, and will fix it as soon as possible if there is a bug found.

Mel-Chen commented:
@fhahn The fix patch: https://github.com/llvm/llvm-project/pull/120395
